index.js

```diff
@@ -0,0 +1,369 @@
+                    }
+
+                    return nextSem(new Error('Specified Key Id not found'));
+                });
+            });
+        }
+    };
+};
+
+/**
+ * Configures the authentication plugin of Hapi
+ * @method implementation
+ * @param  {HapiServer} server           HapiServer we are attaching the auth service to
+ * @param  {Object}     config           Plugin configuration
+ * @param  {Array}      config.keys      List of private keys to sign JWTs with
+ * @param  {String}     config.maxAge    Maximum age for signing JWTs
+ * @param  {String}     config.remoteUri URI of remote keys
+ * @param  {String}     config.timeout   How long to wait for API response
+ * @return {Object}                      Contains an authenticate function to parse each request
+ */
+internals.implementation = (server, config) => {
+    const authConfig = Joi.attempt(config, internals.SCHEMA_KEYS,
+        'Invalid config for hapi-auth-dynamic-jwt plugin');
+    let authObject = {};
+
+    // Local mode
+    if (typeof authConfig.keys === 'object') {
+        authObject = internals.loadLocal(authConfig);
+    } else {
+        authObject = internals.loadRemote(authConfig);
+    }
+
+    return {
+        api: authObject.api,
+
+        /**
+         * Replies with the parsed credentials or an error based on the input from request
+         * @method authenticate
+         * @param  {HapiRequest}   req     Route handler request object
+         * @param  {HapiReply}     reply   Reply interface
+         * @return {Function}              Result of Reply interface
+         */
+        authenticate: (req, reply) => {
+            const parsedHeader = internals.REGEX_BEARER.exec(req.headers.authorization);
+
+            /**
+             * Reply with a unauthorized error message
+             * @method replyError
+             * @param  {String}    message Error message
+             * @return {Promise}
+             */
+            const replyError = message =>
+                reply(Boom.unauthorized(null, 'Bearer', {
+                    error: message
+                }));
+
+            // Ensure we have the header
+            if (!parsedHeader) {
+                return replyError('Missing JWT');
+            }
+
+            const token = parsedHeader[1];
+            const credentials = Jwt.decode(token);
+
+            // Check if it's valid JWT
+            if (!credentials) {
+                return replyError('Invalid JWT format');
+            }
+
+            // Ensure we have the key requested
+            return authObject.getKey(credentials.kid, (keyError, signingKey) => {
+                if (keyError) {
+                    server.log(['auth', 'error', req.name], keyError.toString());
+
+                    return replyError('Unknown JWT public key ID');
+                }
+
+                // Check if the key is expired
+                if (signingKey.expires < Date.now() / 1000) {
```

Comment:
won't in flight requests will still be invalidated and users see error ?

Reply:
JWTs should not be issued using keys that will expire during their lifetime.  See previous comment.